### PR TITLE
Make timer component

### PR DESCRIPTION
**Pull request by chehak-vs**
https://github.com/user-attachments/assets/572cfe30-e3b6-4748-a62b-dc7c766caa68

### CODE REVIEW

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+  const animatedLinearStyle = useAnimatedStyle(() => ({
+    width: `${progress.value * 100}%`,
+    backgroundColor: linearTimerColor,
+  }))
+
+  useEffect(() => {
+    if (isRunning) {
+      const interval = setInterval(() => {
+        setTimeLeft(prev => (prev > 0 ? prev - 1 : 0))
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:96]: we try avoiding states in animation. as well as timeout's/ interval

we can do instead
```  useEffect(() => {
    reanimatedTimer.value = withTiming(0, { duration: duration * 1000, easing: ReanimatedEasing.linear })

    return () => {
      reanimatedTimer.value = 1
    }
  }, [])


  ```

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    setTimeLeft(totalDurationInSeconds)
+    progress.value = 1
+  }, [totalDurationInSeconds, progress])
+
+  const animatedCircleProps = useAnimatedProps(() => ({
+    strokeDashoffset: (1 - progress.value) * 2 * Math.PI * 45,
+  }))
+
+  const animatedLinearStyle = useAnimatedStyle(() => ({
+    width: `${progress.value * 100}%`,
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:89]: don't animated width. Instead use translate and overflow

**Comment by divyanshu-srivastava-vs** [comment]
image


a line is coming in linear progress

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+import styles from './timer-styles'
+
+const AnimatedSvgCircle = Animated.createAnimatedComponent(SvgCircle)
+
+interface ITimerProps {
+  /**buttonStyles: is an optional prop to style the buttons of the timer */
+  buttonStyles?: { container?: ViewStyle; text?: TextStyle }
+  /**circularTimerStrokeColor: is an optional prop that indicates the stroke color of circular timer */
+  circularTimerStrokeColor?: string
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:23]: dont take two props `linearTimerColor` and `circularTimerStrokeColor` instead use a prop name `strokeColor`

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  /**controls: is an optional prop that indicates whether control buttons are visible or not */
+  controls?: boolean
+  /**linearTimerColor: is an optional prop that indicates the color of linear timer */
+  linearTimerColor?: string
+  /**timerType: is a required prop that indicates the type of timer, circular or linear */
+  timerType: TimerPreset
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:29]: timerType -> type

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  /**totalDurationInSeconds: is a required prop that indicates the total number of seconds of timer */
+  totalDurationInSeconds: number
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:31]: duration

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  const {
+    buttonStyles,
+    circularTimerStrokeColor = 'black',
+    controls = false,
+    linearTimerColor = 'black',
+    timerType,
+    totalDurationInSeconds,
+  } = props
+
+  const [timeLeft, setTimeLeft] = useState(totalDurationInSeconds)
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:44]: it will be removed

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            cy="50"
+            r="45"
+            stroke={circularTimerStrokeColor}
+            strokeWidth="10"
+            fill="none"
+            strokeDasharray={`${2 * Math.PI * 45}`}
+            animatedProps={animatedCircleProps}
+          />
+        </Svg>
+        <Text style={styles.timeLeftTextCircular}>{formatTime(timeLeft)}</Text>
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:175]: make a boolean prop which will control whether to show or not for both linear and circular

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    circularTimerStrokeColor = 'black',
+    controls = false,
+    linearTimerColor = 'black',
+    timerType,
+    totalDurationInSeconds,
+  } = props
+
+  const [timeLeft, setTimeLeft] = useState(totalDurationInSeconds)
+  const [isRunning, setIsRunning] = useState(false)
+  const [isPaused, setIsPaused] = useState(false)
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:46]: try using a single state if possible for isRunning and isPaused.

**Review by divyanshu-srivastava-vs — CHANGES_REQUESTED** [review]

**Comment by divyanshu-srivastava-vs** [comment]
image
its not looking as expected

File: AnimatedComponents/App.tsx [hunk 9]
```diff
@@ -103,6 +104,8 @@ const App = () => {
       <AccordionDemo />
 
       <Slider type={SliderType.SingleValueSlider} sliderWidth={300} min={0} max={100} step={5} />
+      <Timer duration={10} type={TimerPreset.Circular} />
+      <Timer duration={10} type={TimerPreset.Linear} controls={true} />
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/App.tsx:108]: <Timer duration={10} type={TimerPreset.Linear} controls />

File: AnimatedComponents/src/components/timer/Timer.tsx [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  const isRunning = useSharedValue(false)
+  const isPaused = useSharedValue(false)
+
+  const handleControl = useCallback(
+    (action: 'start' | 'pause' | 'resume') => {
+      switch (action) {
+        case 'start': {
+          isRunning.value = true
+          isPaused.value = false
+          startAnimation(duration * 1000)
```
Comment (divyanshu-srivastava-vs) [inline AnimatedComponents/src/components/timer/Timer.tsx:63]: 'startAnimation' was used before it was defined.eslint[@typescript-eslint/no-use-before-define](https://typescript-eslint.io/rules/no-use-before-define)

Comment (divyanshu-srivastava-vs) [reply AnimatedComponents/src/components/timer/Timer.tsx:63]: never call a function or variable before its defination

**Comment by divyanshu-srivastava-vs** [comment]
image

image


a line coming in linear in unfilled space. can u please check that?

**Comment by divyanshu-srivastava-vs** [comment]
1. Pause bug (show play/pause at once)
2. duration-> two variable with same name.
3. resolve conflicts

**Comment by divyanshu-srivastava-vs** [comment]
1. Play / pause icons are not at same position. Please check.

**Comment by divyanshu-srivastava-vs** [comment]
reset functionality not working

**Comment by chehak-vs** [comment]
> reset functionality not working

It is working. Please see the below video:
https://github.com/user-attachments/assets/f8e939fb-7d81-4e1b-a75f-99142bc4dd1f

**Pull request merged**